include/gsl/span

```diff
@@ -309,6 +318,12 @@ namespace details
     private:
         index_type size_;
     };
+
+    template <class ElementType, std::ptrdiff_t Extent, std::ptrdiff_t Offset, std::ptrdiff_t Count>
+    struct calculate_subspan_type
+    {
+      using type = span<ElementType, Count != dynamic_extent ? Count : (Extent != dynamic_extent ? Extent - Offset - 1 : Extent)>;
```

Comment:
`Extent - Offset - 1` means that given:
```c++
int data[5]{};
span<int, 5> s = data;
```
`decltype(s.subspan<5>())` is `span<int, -1>` instead of `span<int, 0>`, and `decltype(s.subspan<0>())` is `span<int, 4>` instead of `span<int, 5>`. Surely it should be an invariant that `s.subspan<0>() == s` for a span `s`. This expression should be `Extent - Offset`.

Reply:
See https://github.com/ericniebler/range-v3/blob/master/include/range/v3/span.hpp#L213 =)

Reply:
You are completely right. Nice catch. That will teach me for working at night.

Reply:
In passing...I took a look at the ranges implementation and it doesn't seem to match the intended design for the templated `subspan`: that people can pass a `dynamic_extent` value for the second template parameter to mean "consume everything that remains". (symmetrical to the way the non-template `subspan` works). Is that something you guys chose to do explicitly, or just an outcome of the implementation choice?

Reply:
Oh, and thanks for the bug report!

Reply:
Hahahaha. I thought about the lack of test coverage just after posting the comment....I had a distinct memory you'd run these tests against the implementation...so it's all really my fault (spoiler alert: most things are).

Reply:
And thanks for all the help! You came for the review, stayed for the bug report!